src/engine/interop/specification.md

```diff
@@ -0,0 +1,148 @@
+
+5. Client software **SHOULD** respond with `2: Action not allowed` error if the sync process is in progress.
+
+6. Client software **MUST** respond with `4: Missing resource` error if the parent block is unknown.
+
+### engine_getPayload
+
+#### Parameters
+1. `payloadId`: `QUANTITY`, 64 Bits - Identifier of the payload building process
+
+#### Returns
+`Object|Error` - Either instance of [`ExecutionPayload`](#ExecutionPayload) or an error
+
+#### Specification
+
+1. Given the `payloadId` client software **MUST** respond with the most recent version of the payload that is available in the corresponding building process at the time of receiving the call.
+
+2. The call **MUST** be responded with `4: Missing resource` error if the building process identified by the `payloadId` doesn't exist.
+
+3. Client software **MAY** stop the corresponding building process after serving this call.
+
+### engine_executePayload
+
+#### Parameters
+1. `Object` - Instance of [`ExecutionPayload`](#ExecutionPayload)
+
+#### Returns
+`Object` - Response object:
+1. `status`: `String` - the result of the payload execution:
+  - `VALID` - given payload is valid
+  - `INVALID` - given payload is invalid
+  - `SYNCING` - sync process is in progress
+
+#### Specification
+
+1. Client software **MUST** validate the payload according to the execution environment rule set with modifications to this rule set defined in the [Block Validity](https://eips.ethereum.org/EIPS/eip-3675#block-validity) section of [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) and respond with the validation result.
+
+2. Client software **MUST** defer persisting a valid payload until the corresponding [**`engine_consensusValidated`**](#engine_consensusValidated) message deems the payload valid with respect to the proof-of-stake consensus rules.
+
+3. Client software **MUST** discard the payload if it's deemed invalid.
+
+4. The call **MUST** be responded with `SYNCING` status while the sync process is in progress and thus the execution cannot yet be validated.
+
+5. In the case when the parent block is unknown, client software **MUST** pull the block from the network and take one of the following actions depending on the parent block properties:
+  - If the parent block is a PoW block as per [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) definition, then all missing dependencies of the payload **MUST** be pulled from the network and validated accordingly. The call **MUST** be responded according to the validity of the payload and the chain of its ancestors.
+  - If the parent block is a PoS block as per [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#specification) definition, then the call **MUST** be responded with `SYNCING` status and sync process **SHOULD** be initiated as it is specified in the [merge-sync.md](https://github.com/fjl/p2p-drafts/blob/master/merge-sync/merge-sync.md) document.
```

Comment:
Comments by @djrtwo lost after merging suggestions:

I'm not sure we should specify the specific sync process in this API. (1) there will likely be multiple sync processes supported across different clients and (2) the type of sync (e.g. snap sync vs recursively getting the last N missing blocks and executing them) will be contextually dependent

Additionally, why separate the way this is handled by PoW vs PoS. Pretty much in either mode, the client should decide how to fetch what is missing. Which may be via a state sync protocol or maybe a block sync protocol

Reply:
> I'm not sure we should specify the specific sync process in this API.

Agreed. We better remove it.

> why separate the way this is handled by PoW vs PoS

Suppose the transition process is in progress and the execution client turns into `SYNCING` mode upon receiving the transition block (the child of a terminal block) as it misses the parent of the terminal one. It pulls the parent in a few micros, executes it and verifies the payload and does this in time before attestation needs to be gossiped. Consensus client won't get the status of the transition payload until the next payload is executed and would be responded with `VALID` status. Hence the separation here, in the PoW network execution client rules the sync on its own and it's an opaque process for the consensus client, in the PoS network consensus client becomes more involved in the process.

Though, we might want to avoid handling this corner case taking into account the complexity of implemention.

Reply:
Thoughts on the latter?

Reply:
Okay, I see the difference. 

If parent is PoS block, we can choose whatever sync we want and don't have to stall response.
If parent is PoW block, we must do a recursive block lookup and should not respond until we have an answer.

I'm fine with the text as is as long as we remove the reference to `merge-sync.md`. Pretty much "and a sync process to gather missing blocks and/or state **SHOULD** be initiated"

Reply:
looks good